File: src/android/app/src/main/java/com/dogwalker/app/domain/usecase/auth/RegisterUseCase.kt
package com.dogwalker.app.domain.usecase.auth

// kotlinx.coroutines v1.6.4
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

import com.dogwalker.app.data.repository.AuthRepository
import com.dogwalker.app.domain.model.User

/**
 * Human Tasks:
 * 1. Verify error handling strategy aligns with application-wide error handling policy
 * 2. Configure proper logging for registration events
 * 3. Ensure proper validation rules are implemented for registration fields
 * 4. Review security measures for password handling
 */

/**
 * Use case that encapsulates the business logic for user registration.
 *
 * Requirement addressed: User Management (1.3 Scope/Core Features/User Management)
 * Implements the registration logic by coordinating with the AuthRepository to
 * register new users and store their data securely.
 *
 * @property authRepository Repository that handles authentication-related operations
 */
class RegisterUseCase(
    private val authRepository: AuthRepository
) {
    /**
     * Executes the registration use case with the provided user details.
     *
     * @param name The full name of the user
     * @param email The email address of the user
     * @param password The password for the user account
     * @return User object containing the registered user's details
     * @throws IllegalArgumentException if any of the input parameters are invalid
     * @throws Exception if registration fails
     *
     * Requirement addressed: User Management (1.3 Scope/Core Features/User Management)
     * Implements user registration by validating input parameters and coordinating
     * with the AuthRepository to create new user accounts.
     */
    suspend fun execute(
        name: String,
        email: String,
        password: String
    ): User = withContext(Dispatchers.IO) {
        // Validate input parameters
        require(name.isNotBlank()) { "Name cannot be blank" }
        require(email.isNotBlank()) { "Email cannot be blank" }
        require(password.isNotBlank()) { "Password cannot be blank" }
        
        // Validate length constraints
        require(name.length <= User.MAX_NAME_LENGTH) { 
            "Name cannot exceed ${User.MAX_NAME_LENGTH} characters" 
        }
        require(email.length <= User.MAX_EMAIL_LENGTH) { 
            "Email cannot exceed ${User.MAX_EMAIL_LENGTH} characters" 
        }

        try {
            // Delegate registration to the AuthRepository
            authRepository.register(
                name = name,
                email = email,
                password = password
            )
        } catch (e: Exception) {
            // Re-throw the exception to be handled by the presentation layer
            throw e
        }
    }
}

File: src/ios/DogWalker/Domain/UseCases/Auth/RegisterUseCase.swift
// Foundation framework - Latest
import Foundation

/// RegisterUseCase: Implements the registration process for new users in the Dog Walker application
/// Requirements addressed:
/// - User Management (1.3 Scope/Core Features/User Management): Supports user registration and profile creation
class RegisterUseCase {
    // MARK: - Human Tasks
    /*
    1. Ensure proper validation rules are configured for password strength
    2. Configure email validation patterns if needed
    3. Set up proper error tracking for registration failures
    4. Review and implement proper data sanitization rules
    */
    
    // MARK: - Properties
    
    private let authRepository: AuthRepository
    
    // MARK: - Initialization
    
    /// Initializes the RegisterUseCase with required dependencies
    /// - Parameter authRepository: Repository handling authentication operations
    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }
    
    // MARK: - Public Methods
    
    /// Executes the user registration process
    /// - Parameters:
    ///   - name: User's full name
    ///   - email: User's email address
    ///   - password: User's password
    ///   - phoneNumber: User's phone number
    ///   - role: User's role in the system (e.g., "owner" or "walker")
    /// - Returns: Result containing the registered User or an Error
    func execute(
        name: String,
        email: String,
        password: String,
        phoneNumber: String,
        role: String
    ) -> Result<User, Error> {
        // Validate input parameters
        do {
            try validateRegistrationInput(
                name: name,
                email: email,
                password: password,
                phoneNumber: phoneNumber,
                role: role
            )
        } catch {
            return .failure(error)
        }
        
        // Attempt registration through repository
        return authRepository.register(
            name: name,
            email: email,
            password: password,
            phoneNumber: phoneNumber,
            role: role
        )
    }
    
    // MARK: - Private Methods
    
    /// Validates the registration input parameters
    /// - Parameters:
    ///   - name: User's full name to validate
    ///   - email: User's email to validate
    ///   - password: User's password to validate
    ///   - phoneNumber: User's phone number to validate
    ///   - role: User's role to validate
    /// - Throws: ValidationError if any parameter is invalid
    private func validateRegistrationInput(
        name: String,
        email: String,
        password: String,
        phoneNumber: String,
        role: String
    ) throws {
        // Validate name
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ValidationError.invalidName("Name cannot be empty")
        }
        
        // Validate email format
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        guard emailPredicate.evaluate(with: email) else {
            throw ValidationError.invalidEmail("Invalid email format")
        }
        
        // Validate password strength
        guard password.count >= 8,
              password.rangeOfCharacter(from: .uppercaseLetters) != nil,
              password.rangeOfCharacter(from: .lowercaseLetters) != nil,
              password.rangeOfCharacter(from: .decimalDigits) != nil else {
            throw ValidationError.invalidPassword("Password must be at least 8 characters and contain uppercase, lowercase, and numbers")
        }
        
        // Validate phone number format
        let phoneRegex = "^\\+?[1-9]\\d{1,14}$"
        let phonePredicate = NSPredicate(format: "SELF MATCHES %@", phoneRegex)
        guard phonePredicate.evaluate(with: phoneNumber) else {
            throw ValidationError.invalidPhoneNumber("Invalid phone number format")
        }
        
        // Validate role
        let validRoles = ["owner", "walker"]
        guard validRoles.contains(role.lowercased()) else {
            throw ValidationError.invalidRole("Role must be either 'owner' or 'walker'")
        }
    }
}

// MARK: - ValidationError

/// Enumeration of possible validation errors during registration
private enum ValidationError: LocalizedError {
    case invalidName(String)
    case invalidEmail(String)
    case invalidPassword(String)
    case invalidPhoneNumber(String)
    case invalidRole(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidName(let message),
             .invalidEmail(let message),
             .invalidPassword(let message),
             .invalidPhoneNumber(let message),
             .invalidRole(let message):
            return message
        }
    }
}